rcl/test/rcl/test_publisher_wait_all_ack.cpp

```diff
@@ -0,0 +1,268 @@
+  rcl_ret_t ret;
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  constexpr char topic_name[] = "test_wait_for_all_acked";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  publisher_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  publisher_options.qos.depth = 10000;
+  ret = rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
+  subscription_options.qos.depth = 1;
+  subscription_options.qos.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+
+#define INIT_SUBSCRIPTION(idx) \
+  rcl_subscription_t subscription ## idx = rcl_get_zero_initialized_subscription(); \
+  ret = rcl_subscription_init( \
+    &subscription ## idx, \
+    this->node_ptr, \
+    ts, \
+    topic_name, \
+    &subscription_options); \
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str; \
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT( \
+  { \
+    ret = rcl_subscription_fini(&subscription ## idx, this->node_ptr); \
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str; \
+  });
+
+  INIT_SUBSCRIPTION(1)
+  INIT_SUBSCRIPTION(2)
+  INIT_SUBSCRIPTION(3)
+
+  ASSERT_TRUE(wait_for_established_subscription(&publisher, 10, 100));
+
+  // Prepare 1MB message
+#define ONE_MEGABYTE (1024 * 1024)
+  char test_string[ONE_MEGABYTE];
+  memset(test_string, 'a', ONE_MEGABYTE);
+  test_string[ONE_MEGABYTE - 1] = '\0';
+  test_msgs__msg__Strings msg;
+  test_msgs__msg__Strings__init(&msg);
+  ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg.string_value, test_string));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__msg__Strings__fini(&msg);
+  });
+
+  ret = rcl_publish(&publisher, &msg, nullptr);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription1, context_ptr, 10, 100));
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription2, context_ptr, 10, 100));
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription3, context_ptr, 10, 100));
+
+  int i = 0;
+  for (; i < 500; i++) {
+    ret = rcl_publish(&publisher, &msg, nullptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  ret = rcl_publisher_wait_for_all_acked(
+    &publisher,
+    RCL_MS_TO_NS(500));
+
+  if (is_cyclonedds) {
+    // cyclonedds use sync publish, so above scenario cannot lead to timeout.
+    EXPECT_EQ(RCL_RET_OK, ret);
+  } else {
+    EXPECT_EQ(RCL_RET_TIMEOUT, ret);
+  }
```

Comment:
can we do something like

```cpp
  if (ret != RCL_RET_OK && ret != RCL_RET_TIMOEUT) {
    ASSERT_FAILURE()
  }
```

?

Other rmw implementations can also be configured to use synchronous publishing, so better to do the test independent of the rmw implementation.

Reply:
If modify as you said, rcl test with cyclonedds always report failure in CI.  
Developers have to check what is failed for rcl test.  
Does this failure bring inconvenience for other developer ？

Reply:
> If modify as you said, rcl test with cyclonedds always report failure in CI.

Sorry, but I think that's not the case.
For cyclonedds `ret` will be `RCL_RET_OK`, so `ret != RCL_RET_OK && ret != RCL_RET_TIMOEUT` is `false` and we don't enter the `if()`.
i.e. I want to modify the test so it passes when the function either returns RMW_RET_OK or RMW_RET_TIMEOUT, without listing what rmw implementation is supposed to return which value.

Reply:
Okay. I understand.